### PR TITLE
RFC: deprecate per-module definition of `eval(module, expr)`

### LINKED ISSUE #27093: Define `eval` and `include` consistently in non-bare modules?
When a module is defined generic functions for `eval` and `include` are added.  However, `eval` has two methods `eval([m::Module,] expr)` and `include` only one `include(str)`.  Shouldn't those be symmetric?  I would argue that the `eval(m, expr)` should be dropped; if one really wants to evaluate/include into another module (which should be rare), calling it via the `Core.eval`/`Core.include` is fine (or via `OtherModule.eval(expr)`). 

Removing one of the `eval`-methods would be a breaking change, adding one `include` method wouldn't be.

Edit: note that "my solution" is what the current docs of `eval` state:
```
help?> eval                                                                        
search: eval evalfile @eval @evalpoly bytesavailable readavailable TypeVar         
                                                                                   
  eval([m::Module], expr::Expr)                                                    

  Evaluate an expression in the given module and return the result. Every Module
  (except those defined with baremodule) has its own 1-argument definition of
  eval, which evaluates expressions in that module.
```
So, if no change to eval is made then the doc-string needs updating.

### PR DESCRIPTION
**Pull request by JeffBezanson**
These definitions are silly, since they're identical in every module. Making it slightly less convenient to evaluate in a given module is also a good thing, since you should really think twice before doing that.

fixes #27093

### CODE REVIEW

**Comment by mbauman** [comment]
Does this still allow for `@eval Base …` at the REPL?

**Comment by JeffBezanson** [comment]
Yes; `@eval` is unaffected.

File: src/jlfrontend.scm [hunk 1]
```diff
@@ -131,10 +131,15 @@
...
            (call (core eval) ,name ,x)))
-       (= (call eval m ,x)
-          (block
-           ,loc
-           (call (core eval) m ,x)))
+       (if (&& (call (top isdefined) (core Main) (quote Base))
+               (call (top isdefined) (top Base) (quote @deprecate)))
+           (call eval
+                 (quote
+                  (macrocall (top @deprecate)
```
Comment (Keno) [inline src/jlfrontend.scm]: I'm worried this'll get missed in the deprecation removal. Add a note in deprecated.jl to remove this?

**Comment by JeffBezanson** [comment]
Another thing to consider here is `eval` in `Main` (i.e. at the prompt). Currently it just points to `Core.eval`. If we make this change, we'll probably want to give `Main` its own 1-argument `eval` to match the behavior in other modules (and also deprecate `Core.eval(x)`, which evaluates in `Main`).

**Comment by JeffBezanson** [comment]
Ok I've tried to remove every `eval` method except `Core.eval(mod, ex)` and the per-module `eval(ex)`.

File: base/docs/basedocs.jl [hunk 2]
```diff
@@ -1800,4 +1800,13 @@ The syntax `a.b = c` calls `setproperty!(a, :b, c)`.
 """
 Base.setproperty!
 
+"""
+    eval(expr)
+
+Evaluate an expression in the current module. Every `Module` (except those defined
```
Comment (mauro3) [inline base/docs/basedocs.jl]: Why not "Evaluate an expression in the global scope of the current module." as this is sometimes confusing folks using `eval` inside a function.

**Comment by mauro3** [comment]
If `Core.eval` is the one to use when doing 2-arg calls, the same should go for `include`: i.e. remove the `Base.include(m, p)` method.

If you wanted to make it saver that people do not accidentally evaluate stuff into `Core` (as the `Core.eval` function will get more usage now), you could drop the `Core.eval(e)` method.  So one would have to be very explicit to evaluate into `Core`: `Core.eval(Core, e)`.

File: base/client.jl [hunk 3]
```diff
@@ -180,8 +180,8 @@ function eval_user_input(@nospecialize(ast), show_value::Bool)
                 errcount, lasterr = 0, ()
             else
                 ast = Meta.lower(Main, ast)
-                value = eval(Main, ast)
-                eval(Main, Expr(:body, Expr(:(=), :ans, QuoteNode(value)), Expr(:return, nothing)))
+                value = Main.eval(ast)
+                Main.eval(Expr(:body, Expr(:(=), :ans, QuoteNode(value)), Expr(:return, nothing)))
```
Comment (vtjnash) [inline base/client.jl]: This part of the changes seems strictly wrong. `Main.eval` is a completely random function that's entirely unrelated to the `Core.eval(Main, "")` function this previously called into.

**Comment by fredrikekre** [comment]
Documenter uses this quite a lot, but mostly for doctest. I don't think the CI failures are due to Documenter though, and I don't think the doc build time increased that much from the deprecations here, but see https://github.com/JuliaDocs/Documenter.jl/pull/726

File: base/client.jl (hunk 3 above)
Comment (JeffBezanson) [reply base/client.jl]: Well, it's the `eval` function that I put there, but yes I suppose somebody could overwrite it.

**Comment by JeffBezanson** [comment]
> you could drop the Core.eval(e) method

I did. This previously evaluated in `Main`; I deprecated it.

**Comment by JeffBezanson** [comment]
A possible, even simpler, alternative is to remove all the per-module `eval` functions and use only the 2-argument `eval` exported from `Core`. That would mean `eval(x)` is deprecated to `eval(@__MODULE__, x)`. That seems pretty annoying, but then again I'm not sure how often `eval(...)` is used compared to `@eval`.

**Comment by tkoolen** [comment]
Should there be a `NEWS.md` entry for this?

**Pull request merged**